### PR TITLE
Total Project for Review

### PR DESCRIPTION
**Pull request by Whoistolu**
# Total Project for Review
### On this milestone, we:

- Created and implemented all components in relation to Mission. Below are the details;
- Created a directory for all Redux state slice files.
- Fetched data from the Missions endpoint (https://api.spacexdata.com/v3/missions) when a user navigates to the Missions section.
- Fetched all the rockets data from the RocketsAPI: `https://api.spacexdata.com/v3/rockets`
- Mapped through the data to get the properties values needed id, description, name, and image.
- Once the data were fetched, dispatched an action to store the selected data in Redux store: mission_id, mission_name, description.
- Implemented Missions actions buttons. When a user clicks the "Join Mission" button, an action is dispatched to update the store.
- Implemented the same logic for leaving the mission button.
- Implemented Switch badges for conditional components. Missions that the user has joined already should show a badge "Active Member" instead of the default "NOT A MEMBER" and a button "Leave Mission" instead of the "Join Mission" button (as per design).
- Added styling to my mission table.
- Add my Profile to components to display missions and rockets.

### CODE REVIEW

File: src/components/RocketElement.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import { useSelector, useDispatch } from 'react-redux';
+import { getRockets } from '../redux/rockets/rockets';
+import SingleRocket from './SingleRockets';
+
+const RocketElement = () => {
+  const dispatch = useDispatch();
+  const rockets = useSelector((state) => state.rockets);
+  useEffect(() => {
+    dispatch(getRockets());
+  }, []);
```
Comment (D4nielJ) [inline src/components/RocketElement.js]: - [x] Please make sure that every time a component renders again, it doesn't fetch rockets or missions if they have already been fetched one time.

```
Actions should be dispatch once
Data should not be added to the store on every re-render (i.e. when changing views / using navigation).
```

only-one-fetch

File: src/components/missions.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+function Missions() {
+  const state = useSelector((state) => state.missionReducer);
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    if (state.length === 0) {
+      dispatch(loadMisionsThunk());
+    }
+  }, []);
```
Comment (D4nielJ) [inline src/components/missions.js:15]: - [x] My comment on only fetching data on the first time a component is rendered also applies here.

File: src/components/missions.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  console.log(state);
```
Comment (D4nielJ) [inline src/components/missions.js]: - [x] Let's remove any `console.log` statement that we might have in the code. These are good for debugging, but should not be committed to production code.

File: src/components/rocketsApi.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
+const rocketsUrl = 'https://api.spacexdata.com/v3/rockets';
+
+const fetchApiRockets = async () => {
+  const response = await fetch(rocketsUrl, {
+    method: 'GET',
+  });
+  const result = await response.json();
+  // console.log(result);
```
Comment (D4nielJ) [inline src/components/rocketsApi.js]: - [x] My comment about production code also applies here

File: src/redux/missions/missionReducer.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+            name: item.mission_name,
+            id: item.mission_id,
+            description: item.description,
+            join: false,
+          }));
+          return data;
+        });
+      dispatch(getMissions(req));
+    } catch (e) {
+      console.log(e);
```
Comment (D4nielJ) [inline src/redux/missions/missionReducer.js]: [OPTIONAL] When catching errors you might want to update the state of the application to some `error`, and use this to display some message to the user. Like `There's been an error while loading 'something', you might want to try to reload`

**Review by D4nielJ — CHANGES_REQUESTED** [review]
## Hello @Whoistolu and @luqmanmusah , 

You're doing great! 👏 
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights :fire: 
- Redux and React are well structured. :rocket: 
- Your functionality is ready, just minor tweaks on performance are needed.
- Your UI looks good!


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

## Happy coding and stay awesome! 🚀 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear or you can always find me on slack!

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/components/RocketElement.js (hunk 1 above)
Comment (Whoistolu) [reply src/components/RocketElement.js]: ### Hello @D4nielJ ,
Thank you for your feedback.
The issue has just been fixed. Every time a component renders, it doesn't fetch rockets or missions again.

File: src/components/missions.js (hunk 2 above)
Comment (Whoistolu) [reply src/components/missions.js:15]: Thank you for this.
The issue just got resolved.

File: src/components/missions.js (hunk 3 above)
Comment (Whoistolu) [reply src/components/missions.js]: Thank you for your feedback.
All console.log statements have been removed.

File: src/components/rocketsApi.js (hunk 4 above)
Comment (Whoistolu) [reply src/components/rocketsApi.js]: Thank you for your feedback.
All console.log statements have been removed.

File: src/components/RocketElement.js (hunk 1 above)
Comment (Lameck1) [reply src/components/RocketElement.js]: Done ✔️

File: src/components/missions.js (hunk 2 above)
Comment (Lameck1) [reply src/components/missions.js:15]: Done ✔️

File: src/components/missions.js (hunk 3 above)
Comment (Lameck1) [reply src/components/missions.js]: Done ✔️

File: src/components/rocketsApi.js (hunk 4 above)
Comment (Lameck1) [reply src/components/rocketsApi.js]: Done ✔️

**Review by Lameck1 — APPROVED** [review]
Hi, @Whoistolu & @luqmanmusah,

You've done well addressing concerns in the previous review 👏🏼.
Your project is now complete!  Consider merging this branch :shipit: 

Congratulations! 🎉 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**